test/functional/api/highlight_spec.lua

```diff
@@ -692,6 +692,22 @@ describe('API: get highlight', function()
     api.nvim_set_hl(0, 'Bar', { link = 'Foo', default = true, force = true })
     eq({ link = 'Foo', default = true }, api.nvim_get_hl(0, { name = 'Bar' }))
   end)
+
+  it('can set default link even if link is set', function()
+    api.nvim_set_hl(0, 'From', { link = 'To1' })
+    api.nvim_set_hl(0, 'From', { link = 'To2', default = true })
+    eq({ link = 'To1' }, api.nvim_get_hl(0, { name = 'From' }))
+    command('hi clear')
+    eq({ link = 'To2' }, api.nvim_get_hl(0, { name = 'From' }))
+  end)
+
+  it('does not overwrite default link', function()
+    api.nvim_set_hl(0, 'From', { link = 'To1', default = true })
+    api.nvim_set_hl(0, 'From', { link = 'To2', default = true })
+    eq({ link = 'To1', default = true }, api.nvim_get_hl(0, { name = 'From' }))
```

Comment:
My intuition says this one should be last one wins. I don't know what's _correct_ though.

Reply:
This is the behavior of `:hi def link` so I tried to maintain this behavior. However, I think last-one-wins is better because currently there is no way to update the default link without changing the current link. So if you are fine with the discrepancy between `:hi` and `nvim_set_hl`, I'll change this to last-one-wins.

Reply:
Hmm, might need others to weigh in. My vote would be to change this to last-one-wins since that is more useful.

@clason @justinmk @bfredl @echasnovski thoughts?

Reply:
Ideally, I would like for `nvim_set_hl()` to be as similar to `:highlight` as possible. But since it is already not like that for setting attributes (and this is explicitly documented), this can be approached from a more practical angle.

The only difference this choice can make is when there are multiple calls to set default link for highlight group. Right now, I can not think about any reasonable scenario when this can happen. Default links are usually set by the "provider" of highlight group (be it plugin, syntax file, etc.), anything else should not have to set default link.

So as it *should* not really matter in practice, I'd personally go with preserving compatibility.

Reply:
I don't think ~compatibility~ symmetry is particularly important here. If a user wants to opt-in to the old behaviour (for whatever arcane reason), they always have the option to use `vim.cmd.highlight()`. The last-one-wins behaviour means plugins can reset (or remove) any default highlights they have set up instead of writing them in stone until Nvim is closed.

If we every wanted to get to a point where plugins can update (or reload) without restart (in the far far future), this is something we would need.

Reply:
> Default links are usually set by the "provider" of highlight group (be it plugin, syntax file, etc.), anything else should not have to set default link.

I actually had run into a case where I wanted to override default link only. Sometimes I don't like the default link (say, to `h1`) defined in the plugin's `syntax/some-ft.vim`, and want to override this link. Suppose I use `hi! def link` in `after/syntax/some-ft.vim` to override that default link to `h2`. The fun part is when the colorscheme has its own customization using `hi! link` to `h3`, which I prefer to `h2`. But since syntax scripts run after colorscheme scripts, the link will be overwritten to `h2`, which is suboptimal. For optimial results, `after/syntax/some-ft.vim` should override default link only, without changing the current link.

Reply:
> I actually had run into a case where I wanted to override default link only. Sometimes I don't like the default link (say, to `h1`) defined in the plugin's `syntax/some-ft.vim`, and want to override this link. Suppose I use `hi! def link` in `after/syntax/some-ft.vim` to override that default link to `h2`.

What does overriding **default** link achieve here? If you want some group to be linked to `h2`, use `hi! link group h2`. Default links are for this highlight group providers to have a reasonable default. Anything beyond that is non-default and is indeed "last one wins".

Reply:
> Default links are for this highlight group providers to have a reasonable default. 

The problem is when I think the provider's default is not reasonable.

> What does overriding default link achieve here?

I may want to use other colorschemes that don't have special customization for the group under question. I want to get my version of reasonable default in those colorschemes, without writing per-colorscheme customization myself.

Reply:
That being said, the lack of ability to override the default link (IIRC) can be worked around by setting up a `ColorSchemePre` autocmd in `after/syntax/some-ft.vim`